src/Core/Challenge/Dns/Traits/TopLevelDomainTrait.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+/*
+ * This file is part of the Acme PHP project.
+ *
+ * (c) Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace AcmePhp\Core\Challenge\Dns\Traits;
+
+use LayerShifter\TLDDatabase\Exceptions\ParserException;
+use LayerShifter\TLDExtract\Extract as TLDExtract;
+
+trait TopLevelDomainTrait
+{
+    /**
+     * @param string $domain
+     *
+     * @return string
+     */
+    protected function getTopLevelDomain($domain)
+    {
+        $extract = new TLDExtract();
+        $parse = $extract->parse(str_replace('*.', '', $domain));
+        if (!$parse->isValidDomain()) {
+            throw new ParserException($domain.' is not a valid domain', 1);
```

Comment:
the `ParserException` does not belongs to the AcmePhp namespace. You shouldn't throw this exception here.

Reply:
Actually, I'm trying to fix gandi solver's toplevel not following public suffix listing bug. and are you recommending me to extend `LayerShifter\TLDDatabase\Exceptions\ParserException` into `Acmephp\\*` namespace ?
 https://github.com/acmephp/acmephp/blob/9eeb20a1fd1ed164a76978d965b99ab3afaff5a9/src/Core/Challenge/Dns/GandiSolver.php#L157-L165